unsaved_fast_iterator.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	// We need to ensure that we iterate over saved and unsaved state in order.
+	// The strategy is to sort unsaved nodes, the fast node on disk are already sorted.
+	// Then, we keep a pointer to both the unsaved and saved nodes, and iterate over them in order efficiently.
+	for _, fastNode := range unsavedFastNodeAdditions {
+		if start != nil && bytes.Compare(fastNode.key, start) < 0 {
+			continue
+		}
+
+		if end != nil && bytes.Compare(fastNode.key, end) >= 0 {
+			continue
+		}
+
+		iter.unsavedFastNodesToSort = append(iter.unsavedFastNodesToSort, string(fastNode.key))
+	}
+
+	sort.Slice(iter.unsavedFastNodesToSort, func(i, j int) bool {
+		if ascending{
+			return iter.unsavedFastNodesToSort[i] < iter.unsavedFastNodesToSort[j]
+		} else {
+			return iter.unsavedFastNodesToSort[i] > iter.unsavedFastNodesToSort[j]
+		}
+	})
+
+	if iter.ndb == nil {
+		iter.err = errFastIteratorNilNdbGiven
+		iter.valid = false
+		return iter
+	}
+
+	if iter.unsavedFastNodeAdditions == nil {
+		iter.err = errUnsavedFastIteratorNilAdditionsGiven
+		iter.valid = false
+		return iter
+	}
+
+	if iter.unsavedFastNodeRemovals == nil {
+		iter.err = errUnsavedFastIteratorNilRemovalsGiven
+		iter.valid = false
+		return iter
+	}
+
+	// Move to the first elemenet
+	iter.Next()
+
+	return iter
+}
+
+// Domain implements dbm.Iterator.
+// Maps the underlying nodedb iterator domain, to the 'logical' keys involved.
+func (iter *UnsavedFastIterator) Domain() ([]byte, []byte) {
+	return iter.start, iter.end
+}
+
+// Valid implements dbm.Iterator.
+func (iter *UnsavedFastIterator) Valid() bool {
+	if iter.start != nil && iter.end != nil {
+		if bytes.Compare(iter.end, iter.start) != 1 {
+			return false
+		}
+	}
+
+	return iter.fastIterator.Valid() || iter.nextUnsavedNodeIdx < len(iter.unsavedFastNodesToSort) || iter.nextKey != nil || iter.nextVal != nil
+}
+
+// Key implements dbm.Iterator
+func (iter *UnsavedFastIterator) Key() []byte {
+	return iter.nextKey
+}
+
+// Value implements dbm.Iterator
+func (iter *UnsavedFastIterator) Value() []byte {
+	return iter.nextVal
+}
+
+// Next implements dbm.Iterator
+func (iter *UnsavedFastIterator) Next() {
+	if iter.ndb == nil {
+		iter.err = errFastIteratorNilNdbGiven
+		iter.valid = false
+		return
+	}
+
+	if iter.fastIterator.Valid() && iter.nextUnsavedNodeIdx < len(iter.unsavedFastNodesToSort) {
+		diskKeyStr := string(iter.fastIterator.Key())
+
+		if iter.unsavedFastNodeRemovals[diskKeyStr] != nil {
+			// If next fast node from disk is to be removed, skip it.
+			iter.fastIterator.Next()
+			iter.Next()
+			return
+		}
+
+		nextUnsavedKey := iter.unsavedFastNodesToSort[iter.nextUnsavedNodeIdx]
+		nextUnsavedNode := iter.unsavedFastNodeAdditions[nextUnsavedKey]
+
+		var isUnsavedNext bool
+		if iter.ascending {
+			isUnsavedNext = bytes.Compare([]byte(diskKeyStr), []byte(nextUnsavedKey)) >= 0
+		} else {
+			isUnsavedNext = bytes.Compare([]byte(diskKeyStr), []byte(nextUnsavedKey)) <= 0
```

Comment:
In a future PR, we can optimize the heap allocations here by making an `UnsafeStringToBytes` function,  e.g. defined here: https://github.com/cosmos/cosmos-sdk/blob/master/internal/conv/string.go#L8-L18

Reply:
Or alternatively this for the initial []byte to string conversion: https://github.com/cosmos/cosmos-sdk/blob/ebddeee36095dcf93353a9b6d2d59b444a077172/internal/conv/string.go#L20-L26

Reply:
I think these conversions are redundant. We can compare the strings directly. Will get that fixed

Reply:
Why are these conversions more efficient though?

Reply:
Basically every type cast between string and []byte requires re-allocating that data on heap. This is because golang guarantees that if you make a string of some []bytes, that strings bytes can never be altered. (e.g. by a prior reference to the old byte slice)

Similarly, if you cast a byte slice to a string, since you can still edit the old byte slice it allocates it on heap.

So these unsafe conversions allow you to do the conversion without heap additions. It requires the caller to guarantee that it doesn't edit the []byte.

Reply:
addressed and thanks for the explanation